cpp/oneapi/dal/algo/svm/backend/gpu/smo_solver.hpp

```diff
@@ -0,0 +1,280 @@
+
+    auto solve_event = queue.submit([&](sycl::handler& cgh) {
+        cgh.depends_on(deps);
+        local_accessor_rw_t<Float> local_kernel_values(ws_count, cgh);
+        local_accessor_rw_t<Float> objective_func(ws_count, cgh);
+        local_accessor_rw_t<enumerate_value<Float>> sg_cache(max_sg_size, cgh);
+        local_accessor_rw_t<local_variables<Float>> local_vars(1, cgh);
+
+        cgh.parallel_for(nd_range, [=](sycl::nd_item<1> item) {
+            const std::uint32_t i = item.get_local_id(0);
+
+            const std::uint32_t ws_index = ws_indices_ptr[i];
+
+            Float f_i = f_ptr[ws_index];
+            Float alpha_i = alpha_ptr[ws_index];
+            const Float old_alpha_i = alpha_i;
+            const Float labels_i = labels_ptr[ws_index];
+
+            std::uint32_t b_i = 0;
+            std::uint32_t b_j = 0;
+
+            Float* local_kernel_values_ptr = local_kernel_values.get_pointer().get();
+            Float* objective_func_ptr = objective_func.get_pointer().get();
+            enumerate_value<Float>* sg_cache_ptr = sg_cache.get_pointer().get();
+            local_variables<Float>* local_vars_ptr = local_vars.get_pointer().get();
+
+            local_kernel_values_ptr[i] = kernel_values_ptr[i * row_count + ws_index];
+            item.barrier(sycl::access::fence_space::local_space);
+
+            std::uint32_t inner_iter = 0;
+            for (; inner_iter < max_inner_iter; ++inner_iter) {
+                /* m(alpha) = min(grad[i]): i belongs to I_UP (alpha) */
+                objective_func_ptr[i] = is_upper_edge<Float>(labels_i, alpha_i, C) ? -f_i : fp_min;
+
+                /* Find i index of the working set (b_i) */
+                reduce_arg_max(item,
+                               objective_func_ptr,
+                               sg_cache_ptr,
+                               local_vars_ptr[0].max_val_ind);
+                b_i = local_vars_ptr[0].max_val_ind.index;
+                const Float ma = -local_vars_ptr[0].max_val_ind.value;
+
+                /* max_f(alpha) = max(grad[i]): i belongs to i_low (alpha)  */
+                objective_func_ptr[i] = is_lower_edge<Float>(labels_i, alpha_i, C) ? f_i : fp_min;
+
+                /* Find max gradient */
+                reduce_arg_max(item,
+                               objective_func_ptr,
+                               sg_cache_ptr,
+                               local_vars_ptr[0].max_val_ind);
+
+                if (i == 0) {
+                    const Float max_f = local_vars_ptr[0].max_val_ind.value;
+
+                    /* for condition check: m(alpha) >= max_f */
+                    local_vars_ptr[0].local_diff = max_f - ma;
+                    if (inner_iter == 0) {
+                        local_vars_ptr[0].local_eps =
+                            sycl::fmax(eps, local_vars_ptr[0].local_diff * Float(1e-1));
+                        f_diff_ptr[0] = local_vars_ptr[0].local_diff;
```

Comment:
I here missed a mistake in due time. It is always correct to compute the difference, not only at the first iteration. We fixed  in CPU for 2021.4 release. See [here](https://github.com/oneapi-src/oneDAL/blob/master/cpp/daal/src/algorithms/svm/svm_train_thunder_impl.i#L467)
You need to move diff before if:
```c++
local_vars_ptr[0].local_diff = max_f - ma;
if (inner_iter == 0) {
...
}
item.barrier(sycl::access::fence_space::local_space);
if (local_vars_ptr[0].local_diff < local_vars_ptr[0].local_eps) {
    break;
}